sys/net/link_layer/ng_nomac/ng_nomac.c

```diff
@@ -0,0 +1,153 @@
+    int res;
+    msg_t msg, reply, msg_queue[NG_NOMAC_MSG_QUEUE_SIZE];
+
+    /* setup the MAC layers message queue */
+    msg_init_queue(msg_queue, NG_NOMAC_MSG_QUEUE_SIZE);
+    /* save the PID to the device descriptor */
+    dev->mac_pid = thread_getpid();
+    /* register the event callback with the device driver */
+    dev->driver->add_event_callback(dev, _event_cb);
+
+    /* start the event loop */
+    while (1) {
+        DEBUG("nomac: waiting for incoming messages\n");
+        msg_receive(&msg);
+        /* dispatch NETDEV and NETAPI messages */
+        switch (msg.type) {
+            case NG_NETDEV_MSG_EVENT_TYPE:
+                DEBUG("nomac: NG_NETDEV_MSG_EVENT_TYPE received");
+                dev->driver->isr_event(dev, msg.content.value);
+                break;
+            case NG_NETAPI_MSG_TYPE_SND:
+                DEBUG("nomac: NG_NETAPI_MSG_TYPE_SND received");
+                dev->driver->send_data(dev, (ng_pktsnip_t *)msg.content.ptr);
+                break;
+            case NG_NETAPI_MSG_TYPE_SETOPT:
+                /* TODO: filter out MAC layer options -> for now forward
+                         everything to the device driver */
+                DEBUG("nomac: NG_NETAPI_MSG_TYPE_SETOPT received");
+                /* read incoming options */
+                opt = (ng_netapi_opt_t *)msg.content.ptr;
+                /* set option for device driver */
+                res = dev->driver->set(dev, opt->opt, opt->data, opt->data_len);
+                /* send reply to calling thread */
+                reply.type = NG_NETAPI_MSG_TYPE_ACK;
+                reply.content.value = (uint32_t)res;
+                msg_reply(&msg, &reply);
+                break;
+            case NG_NETAPI_MSG_TYPE_GETOPT:
+                /* TODO: filter out MAC layer options -> for now forward
+                         everything to the device driver */
+                DEBUG("nomac: NG_NETAPI_MSG_TYPE_GETOPT received");
+                /* read incoming options */
+                opt = (ng_netapi_opt_t *)msg.content.ptr;
+                /* get option from device driver */
+                res = dev->driver->get(dev, opt->opt, opt->data,
+                                       (size_t*)(&(opt->data_len)));
+                /* send reply to calling thread */
+                reply.type = NG_NETAPI_MSG_TYPE_ACK;
+                reply.content.value = (uint32_t)res;
+                msg_reply(&msg, &reply);
+                break;
+        }
+    }
+    /* never reached */
+    return NULL;
+}
+
+kernel_pid_t nomac_init(char *stack, int stacksize, char priority,
+                        const char *name, ng_netdev_t *dev)
+{
+    kernel_pid_t res;
+
+    /* check if given netdev device is defined */
+    if (dev == NULL) {
+        return -ENODEV;
+    }
+    /* create new NOMAC thread */
+    res = thread_create(stack, stacksize, priority, CREATE_STACKTEST,
+                         _nomac_thread, (void *)dev, name);
```

Comment:
I'm not sure about this, but for the sake of discussion: shouldn't there be a

```C
ng_netif_add(res);
```

in here? Or do we want to do this in an as of yet to specify bootstrapping place?